src/mlpack/methods/ann/rbm_impl.hpp

```diff
@@ -0,0 +1,237 @@
+        FreeEnergy(std::move(negativeSamples)));
+  }
+  else
+  {
+    if (persistence)
+    {
+      return rbmPolicy.Evaluate(predictors, i, predictors.n_cols);
+    }
+    else
+    {
+      // Mean Squared Error
+      rbmPolicy.SampleHidden(std::move(predictors.col(i)),
+          std::move(hiddenReconstruction));
+      rbmPolicy.SampleVisible(std::move(hiddenReconstruction),
+          std::move(visibleReconstruction));
+      return arma::accu(arma::pow(visibleReconstruction - predictors.col(i),
+          2)) / hiddenReconstruction.n_rows;
+    }
+  }
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+void RBM<InitializationRuleType, RBMPolicy>::SampleHidden(
+    arma::Mat<ElemType>&& input,
+    arma::Mat<ElemType>&& output)
+{
+  rbmPolicy.SampleHidden(std::move(input), std::move(output));
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+void RBM<InitializationRuleType, RBMPolicy>::SampleVisible(
+    arma::Mat<ElemType>&& input,
+    arma::Mat<ElemType>&& output)
+{
+  rbmPolicy.SampleVisible(std::move(input), std::move(output));
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+void RBM<InitializationRuleType, RBMPolicy>::Gibbs(
+    arma::Mat<ElemType>&& input,
+    arma::Mat<ElemType>&& output,
+    size_t steps)
+{
+  steps = (steps == SIZE_MAX) ? this-> numSteps: steps;
+
+  if (persistence && !state.is_empty())
+  {
+    SampleHidden(std::move(state), std::move(gibbsTemporary));
+    SampleVisible(std::move(gibbsTemporary), std::move(output));
+  }
+  else
+  {
+    SampleHidden(std::move(input), std::move(gibbsTemporary));
+    SampleVisible(std::move(gibbsTemporary), std::move(output));
+  }
+
+  for (size_t j = 1; j < steps; j++)
+  {
+    SampleHidden(std::move(output), std::move(gibbsTemporary));
+    SampleVisible(std::move(gibbsTemporary), std::move(output));
+  }
+  if (persistence)
+    state = output;
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+void RBM<InitializationRuleType, RBMPolicy>::Gradient(
+    arma::Mat<ElemType>& /*parameters*/,
+    const size_t input,
+    arma::Mat<ElemType>& output,
+    const size_t /*batchSize*/)
+{
+  positiveGradient.zeros();
+  // Collect the negative samples
+  rbmPolicy.PositivePhase(std::move(predictors.col(input)),
+      std::move(positiveGradient));
+
+  negativeGradient.zeros();
+
+  for (size_t i = 0; i < negSteps; i++)
+  {
+    Gibbs(std::move(predictors.col(input)), std::move(negativeSamples));
+    rbmPolicy.NegativePhase(std::move(negativeSamples),
+        std::move(tempNegativeGradient));
+
+    negativeGradient += tempNegativeGradient;
+  }
+  output = ((negativeGradient / negSteps) - positiveGradient);
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+void RBM<InitializationRuleType, RBMPolicy>::Shuffle()
+{
+  predictors = predictors.cols(arma::shuffle(arma::linspace<arma::uvec>(0,
+      predictors.n_cols - 1,
+      predictors.n_cols)));
+}
```

Comment:
@lozhnikov I'm not confident if this is the correct implementation. Please advise.

Reply:
Looks like the implementation is correct.